### PR TITLE
Make it possible to dismiss a snackbar manually

### LINKED ISSUE #463: you should be able to manually swipe a snack bar off-screen

### PR DESCRIPTION
**Pull request by abarth**
Fixes #463

### CODE REVIEW

**Comment by abarth** [comment]
@HansMuller

File: packages/flutter/lib/src/widgets/dismissable.dart [hunk 1]
```diff
@@ -64,7 +64,6 @@ class Dismissable extends StatefulWidget {
     this.onDismissed,
     this.direction: DismissDirection.horizontal
   }) : super(key: key) {
-    assert(key != null);
```
Comment (abarth) [inline packages/flutter/lib/src/widgets/dismissable.dart]: I'm especially not sure about this part of the patch.

File: packages/flutter/lib/src/material/scaffold.dart [hunk 2]
```diff
@@ -304,6 +304,17 @@ class ScaffoldState extends State<Scaffold> {
     }
   }
 
+  void removeCurrentSnackBar() {
+    if (_snackBars.isEmpty)
+      return;
+    Completer<Null> competer = _snackBars.first._completer;
```
Comment (HansMuller) [inline packages/flutter/lib/src/material/scaffold.dart]: I think you mean 'completer' here

File: packages/flutter/lib/src/widgets/dismissable.dart (hunk 1 above)
Comment (HansMuller) [inline packages/flutter/lib/src/widgets/dismissable.dart]: I think removing this assert is OK but a little risky. It stems from: https://github.com/flutter/flutter/pull/2467#discussion_r55267809

Basically it was too hard to write an assert that detected when a missing key was needed and explained as much in a way that a developer could understand.  So we always required a key.  

If you fail to provide a key when it's needed, the failure mode is difficult to diagnose.

**Comment by HansMuller** [comment]
LGTM

File: packages/flutter/lib/src/material/scaffold.dart (hunk 2 above)
Comment (abarth) [inline packages/flutter/lib/src/material/scaffold.dart]: Fixed.

File: packages/flutter/lib/src/widgets/dismissable.dart (hunk 1 above)
Comment (abarth) [inline packages/flutter/lib/src/widgets/dismissable.dart]: Ok, I left the assert in and added a key to the use in snackbar.

File: packages/flutter/lib/src/material/snack_bar.dart [hunk 3]
```diff
@@ -118,23 +119,30 @@ class SnackBar extends StatelessWidget {
...
-                ),
-                child: new FadeTransition(
-                  opacity: fadeAnimation,
-                  child: new Row(
-                    children: children,
-                    crossAxisAlignment: CrossAxisAlignment.center
+          child: new Dismissable(
+            key: new Key('dismissable'),
+            direction: DismissDirection.down,
+            onDismissed: (DismissDirection direction) {
```
Comment (Hixie) [inline packages/flutter/lib/src/material/snack_bar.dart]: dismissable is going to try to shrink the width after the snack bar has slid down. That seems a bit weird. Maybe we should have a variant of dismissable that doesn't do that?

**Comment by Hixie** [comment]
test?

**Comment by abarth** [comment]
Test added in https://github.com/flutter/flutter/pull/3039

**Pull request merged**